### PR TITLE
LLM-1513: Auto-migrate Claude Code and OpenCode setups on first run

### PR DESCRIPTION
**Pull request by krzysztofreczek**
Closes [LLM-1513](https://castai.atlassian.net/browse/LLM-1513).

## What this does

On first run, Kimchi detects existing Claude Code and OpenCode setups and offers to import their MCP servers and commands — so users keep their existing tools without manual reconfiguration.

## User journey

When an existing setup is found, the wizard shows what's available:

```
┌  Claude Code + OpenCode configuration found
│
│  MCP servers: filesystem, github, ripgrep
│  Claude Code skills: 4 in ~/.claude/skills
│  Claude Code commands: 12 in ~/.claude/commands
│  OpenCode skills: 2 in ~/.config/opencode/skills
│
◇  Migrate MCP servers to Kimchi?
│  ● Migrate now
│  ○ Skip this time
│  ○ Never ask again
```

Choosing **Migrate now** opens a per-server selection (all pre-checked):

```
◆  Select MCP servers to migrate (a: toggle all):
│  ◼ filesystem  (Claude Code)
│  ◼ github      (Claude Code)
│  ◻ ripgrep     (OpenCode)
└
```

Then a per-command selection for prompt migration:

```
◆  Select commands to migrate as prompts (a: toggle all):
│  ◼ mr          (Claude Code)
│  ◼ review      (Claude Code)
│  ◼ session-end (Claude Code)
└
```

Selected commands are copied to `~/.config/kimchi/harness/prompts/` with frontmatter added. Subdirectory files (e.g. `reference/react.md`) are migrated automatically alongside selected top-level commands, preserving the original directory structure.

**Skip this time** dismisses until next run. **Never ask again** suppresses permanently.

Migration is only marked complete after all phases finish — if the user interrupts mid-flow, the wizard re-runs next time.

## Demo

[LLM-1513]: https://castai.atlassian.net/browse/LLM-1513?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---
### What changed
Added an automatic migration wizard that detects existing **Claude Code** and **OpenCode** installations on first run, offering to migrate their MCP servers, user skills, and custom commands into Kimchi's configuration. The wizard presents a unified interface showing discovered configurations from all supported agents and allows selective migration of resources.

### Why
Reduces friction for users switching from other coding agents by automatically importing their existing tool configurations and custom prompts rather than requiring manual setup.

### Key changes
- Added `src/agent-discovery/` framework with pluggable `AgentDefinition` interface for detecting third-party agent configurations
- Implemented `claude-code.ts` agent definition supporting `~/.claude.json` (top-level and per-project `mcpServers`) and `~/.claude/skills/` discovery
- Implemented `opencode.ts` agent definition supporting both modern (`mcp` block with `type: "local" | "remote"`) and legacy (`mcpServers` block with `type: "stdio" | "sse"`) schemas, plus `$OPENCODE_CONFIG` environment override and JSONC parsing
- Added `engine.ts` with conflict resolution logic: earlier config files win over later ones, Claude Code wins over OpenCode, and existing Kimchi `mcp.json` entries always win
- Added `jsonc.ts` minimal parser to handle OpenCode's commented configuration files
- Modified `setup-wizard.ts` to merge discoveries from all agents, prompt for selective MCP server migration, and migrate command files (`.md`) from agent commands directories to Kimchi prompts with generated frontmatter
- Updated `config.ts` with `buildSkillPathOptions()` to dynamically include discovered skill directories from other agents in the skills selection UI
- Removed `src/cc-discovery.ts` (functionality superseded by new framework)
- Added connection retry logic in `mcp-adapter/init.ts` for failed non-OAuth MCP servers

### Impact
- Migration is idempotent: re-running the wizard won't overwrite existing Kimchi MCP server entries or previously migrated prompts
- Users can permanently skip migration via "Never ask again", stored in `~/.config/kimchi/config.json` as `migrationState: "skip-forever"`
- Adding support for additional agents (Cursor, Cline, etc.) requires only implementing `AgentDefinition` and appending to `AGENT_DEFINITIONS`
- Configuration parsing failures emit warnings but don't block the wizard; missing files are silently ignored

---
### What changed
Adds automated migration support for **OpenCode** alongside existing Claude Code support. The first-run wizard now detects both agents, offers selective migration of MCP servers, and converts command files (`.md` files in `~/.claude/commands/` or `~/.config/opencode/commands/`) into Kimchi prompts. Introduces a pluggable `AgentDefinition` framework for adding future agent detectors.

### Why
Users switching from Claude Code or OpenCode need a seamless way to import their existing MCP server configurations, user skills, and custom commands without manual file editing.

### Key changes
- **`src/agent-discovery/`** — New framework with `AgentDefinition` interface for pluggable agent detection and `discoverAgent()` engine supporting JSON/JSONC parsing, first-writer-wins merging, and graceful error handling for corrupted files.
- **`src/agent-discovery/agents/opencode.ts`** — Implements OpenCode discovery supporting both modern (`mcp` block) and legacy Go-binary (`mcpServers` block) config schemas, plus JSONC comment stripping.
- **`src/agent-discovery/agents/claude-code.ts`** — Refactors Claude Code discovery into the `AgentDefinition` format, merging project-level `mcpServers` over top-level entries.
- **`src/setup-wizard.ts`** — Extends migration UI to list discovered servers from all agents with multi-select, migrates command files to prompts preserving subdirectory structure and frontmatter, and gates prompts on `migrationState: "skip-forever"` in `~/.config/kimchi/config.json`.
- **`src/extensions/mcp-adapter/init.ts`** — Adds retry logic for transient initialization errors (`EBUSY`, `ECONNREFUSED`, `ETIMEDOUT`, npm lock contention) to improve reliability on first run.
- **`src/config.ts`** — Adds `buildSkillPathOptions()` to dynamically include discovered agent skill directories in the default paths list.
- **Removed** `src/cc-discovery.ts` (functionality superseded by modular agent definitions).

### Impact
- **Breaking**: None; existing Kimchi MCP config entries always win on name collisions, making re-runs idempotent and safe.
- **UX**: Users with existing Claude Code or OpenCode installations will see a one-shot migration prompt on next launch; selecting "Never ask again" persists to `config.json` and suppresses future prompts.
- **Reliability**: MCP server initialization now retries transient failures once before surfacing errors.

### CODE REVIEW

File: src/oc-discovery.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		entry.headers = raw.headers
+		if (raw.url && hasBearerAuthorizationHeader(raw.headers)) {
+			entry.auth = "bearer"
+		}
+	}
+
+	return entry
+}
+
+function normaliseLegacyEnv(env: unknown): Record<string, string> | undefined {
```
[the remark this answers is not in the document]
Comment (krzysztofreczek) [reply src/oc-discovery.ts]: Fixed

File: src/oc-discovery.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	if (raw.command !== undefined) {
+		if (!Array.isArray(raw.command) || raw.command.length === 0) {
+			return undefined
+		}
+		entry.command = raw.command[0]
+		if (raw.command.length > 1) {
+			entry.args = raw.command.slice(1)
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (krzysztofreczek) [reply src/oc-discovery.ts]: Fixed

**Review by castabar — APPROVED** [review]
# PR Review: LLM-1513 — Auto-migrate Claude Code and OpenCode setups on first run

## Overall Assessment: ✅ Approve

Well-structured feature with good architecture, comprehensive tests (91 pass), and clear separation of concerns. The pluggable `AgentDefinition`/`AGENT_DEFINITIONS` design makes extending to Cursor/Cline/Aider trivial.

---

## ✅ What’s strong

| Area | Observation |
|------|-------------|
| **Architecture** | Pluggable `AgentDefinition` interface with `extractServerSources`/`transformServer` handles dual-schema agents (OpenCode modern + legacy) cleanly. Great forward-compatibility. |
| **Defensive coding** | `hasBearerAuthorizationHeader` hardened against `null`/`undefined`/`Array`/`primitive` headers. `ingest` skips malformed entries silently. |
| **Config merging** | First-writer-wins at the server-name level across files and agents. Matches documented conflict resolution and is idempotent. |
| **Tests** | 91 tests across engine, claude-code, opencode, and setup-wizard. Edge cases (corrupt JSON, ENOENT, null headers, empty dirs) are covered. |
| **User experience** | Per-server multiselect with origin hints. Directory structure preserved for migrated commands. "Never ask again" persisted. |
| **Atomic writes** | `writeMcpServers` uses temp-file + `renameSync` for atomicity. Good. |

---

## ⚠️ Comments (non-blocking)

### 1. `runCommandsMigrationPhase` — Ctrl+C mid-flow permanently marks migration done
**Location:** `src/setup-wizard.ts` ~335

After a user selects **"Migrate now"** and completes MCP server selection, if they hit Ctrl+C during the command multiselect, `runCommandsMigrationPhase` returns `0`, but `migrationRan = true` is still set unconditionally:

```typescript
const commandsCopied = await runCommandsMigrationPhase(merged.agents)
// migrationRan = true ← always set, even on cancel
```

This permanently marks `migrationState = "done"`, so the user never gets re-prompted for command migration. This is inconsistent with MCP-select cancellation behaviour, which returns `skip-once` so the wizard re-runs.

**Suggested fix:** propagate the cancel (or return a sentinel) and gate `migrationRan`:

```typescript
// In runCommandsMigrationPhase:
if (clack.isCancel(chosen)) return "cancelled"
// ...

// In runSetupWizard:
if (commandsCopied === "cancelled") {
    migrationRan = false  // don't mark done; re-run next time
} else {
    migrationRan = true
}
```

### 2. MCP connection retry is overly broad
**Location:** `src/extensions/mcp-adapter/init.ts` ~115

All non-OAuth failures are retried sequentially. This adds latency for fundamentally broken servers (missing binary, wrong URL, auth failure). Consider narrowing to transient errors (e.g. `EBUSY`, `ECONNREFUSED`, timeout, npm lock contention) rather than a blanket retry.

### 3. `FRONTMATTER_REGEX` edge case — EOF without trailing newline
**Location:** `src/setup-wizard.ts` ~170

The regex requires `\r?\n` before the closing `---`. If a file ends with `---` at EOF with no trailing newline, the frontmatter won’t match and the entire file gets treated as body. Very rare in practice, but a simple `?` on the trailing newline group would close it.

### 4. `>` (folded) YAML description syntax
**Location:** `src/setup-wizard.ts` ~185

`extractExistingFrontmatter` handles `|` (literal) but not `>` (folded). Claude Code commands with folded multiline descriptions would not preserve their description. Minor; most commands use inline strings.

---

## ❗ Pre-existing (not from this PR)

`pnpm run lint` fails in `src/extensions/improve/index.ts` due to a trailing newline. `git diff HEAD~6` confirms this file is untouched by this PR; the lint failure is pre-existing on the branch.

```
src/extensions/improve/index.ts format
× Formatter would have printed the following content:
```

This will block merge unless either (a) `master` is already clean or (b) you fix it in a preliminary commit.

---

## Summary

Ready to merge. The command-migration cancel state issue (item 1) should be tracked as a follow-up if not addressed here.
